### DISCUSSION TITLE
Exposing telnet functions in Harbor http endpoints?

### DISCUSSION THREAD

**Discussion by Ziphoid** [savonet/liquidsoap#3963] [Q&A]
Hi! I might be a bit blind or so, but I can't really seem to figure out if there's a way to do this. I'm building an external site engine on top of Liquidsoap and would _REALLY_ want to utilise some of the functionality the telnet interface gives, but can't find out how I could do this. Even though you have a lot of documentation available, it's not always clear on how to do things. :)

Primarily, what I'm looking for is a way to http-wrap the calls concering `request`, as in:
> request.alive
request.all
request.metadata <rid>
request.on_air
request.resolving
request.trace <rid>

For now, I've solved it by wrapping shell calls using `nc` like `echo "request.on_air" | nc -q 0 localhost 1234` which works but still feels like a fugly hack if there are better ways to do it. :) Any hints would be really appreciated.

**Comment by vitoyucepi** [comment]
Hi @Ziphoid,

As a starting point, I recommend reading this article https://www.liquidsoap.info/doc-2.2.4/harbor_http.html. It has most of the information you need to try to write a simple http handler.
Have you tried anything yet?

Here's my example for `request.on_air`:
```ruby
# Lets imagine that request.on_air is a function that returns bool
def on_air_handler(_)
  harbor.http.response(data="#{request.on_air()}")
end

harbor.http.register.simple(port=8080, method="GET", "/on_air", on_air_handler)
```

**Reply by Ziphoid** [reply]
Heya! Yes, I've already written a couple of http responses without a problem. What I've failed to manage is to find the content that the telnet commands output. Ie, the current queue and its metadata. I guess it's only a matter of passing the right data at the right moment, but I simply can't find the correct one. I'm currently out spelunking in the real world but will surely dig deeper later tonight. :) Thanx for the reply so far!

**Reply by toots** [reply]
Hi @Ziphoid ! Most if not all the scripted telnet function are here: https://github.com/savonet/liquidsoap/blob/main/src/libs/server.liq

Some other are internal (i.e. written in compile code) but these should be a good start.

Generally speaking, the information you need should be available as methods. For instace, for `request.dynamic`, you have the following methods: https://www.liquidsoap.info/doc-2.2.4/reference.html#request.dynamic
* `queue`
* `set_queue`
* `skip`
etc..

However, for the request commands you mentioned, I am not sure that those are in fact available beside telnet -- those are really old now! However, I'm about to to a pass of cleanup/upgrade of the request code so I'll definitely keep this in mind!

**Reply by Ziphoid** [reply]
Thanx for these pointers! It seems like I'm able to somehow get hold of the data I need, but my knowledge about the LS language just isn't good enough, so I guess I should take a step back and make myself understand what I'm really doing here. :D For now, my faux telnet bridge will have to do while I learn LS scripting a bit more and can rewrite the scripts I have so that they make sense. The reason why I want to be able to reach this data in the first place is that my external engine keeps track of a lot more metadata than I actually inject into LS, so I need to be able to make sure that everything is going as planned. I have noticed that there sometimes are some minor hiccups causing the queue to be a bit out of order, which is why it would help to make sure that the data I have as current in the external system fits the data in the actual LS instance.

Also, I just must underline the fact that I'm happy I bumped into such a welcoming community. Even though the whole scripting experience pretty much contradicts my previous knowledge, the response when actually asking questions have been great, so keep that up! :D 

The thing I'm missing right now would bw some real-world examples of a whole radio flow - preferably including some audio processing chain as well, since I can't get the LADSPA thingy to work at all - but I guess that things will get easier once I get the hang of how it all actually works. It's always so hard to know a bit too much to be a complete novice but way too little to actually know what I'm doing and at the same time be impatient enough to want progress ASAP so I kinda skip through things a bit too fast. :P

**Comment by danbo** [comment]
I'm a bit surprised that request.all is not exposed exposed programmatically /  only over telnet - without this I don't see how I can report back to a UI what is actually queued up to play next.

Is there a plan to do this? Is it not there because of #4093?

**Reply by toots** [reply]
Hi  @danbo. 

I just added `request.all` to the API. It'll come out with the next `rolling-release-v2.3.x`

It should be noted that, to fetch what's coming next, it is recommended to get the data directly from the source, be it `request.dynamic` and etc and to fetch the metadata currently being played directly from the  corresponding output.

**Reply by danbo** [reply]
Thank you. I suppose per your answer in #4093, it won't be as helpful to rely on in a UI, but perhaps still useful when debugging in code vs telnet.
